pkg/engine/constraints.go

```diff
@@ -0,0 +1,127 @@
+package engine
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/klothoplatform/klotho/pkg/core"
+	"github.com/klothoplatform/klotho/pkg/graph"
+	"gopkg.in/yaml.v3"
+)
+
+type (
+	Constraint interface {
+		Level() ConstraintLevel
+		IsSatisfied(dag *core.ResourceGraph) bool
+		CanApply() bool
+		Apply(dag *core.ResourceGraph) error
+	}
```

Comment:
Second comment on `Apply(dag)` (I'm replying on a second line so that we can keep the threads separate): this interface doesn't really let the user see if two constraints are individually possible, but impossible when combined. They'll see the error, but not the real reason.

For example, let's say I have a graph `APIGW -> Compute`, and I say "make the compute a serverless EC2". That's impossible, by definition; but I'm guessing we would turn this into two constraints: "compute must be serverless" and "compute must be EC2", and then apply them in some order. I think the user would get back an error saying either "compute_01 cannot be serverless because it has type EC2" or "compute_01 cannot be EC2 because it is serverless" (or maybe even "compute_01 cannot be EC2 because it is Lambda", depending on how specific `Apply(~)`'s actions get), but the error I'd really want is something like "serverless and EC2 are mutually exclusive".

Reply:
yeah the interface isnt thought through yet, but also will not be publicly visible. Eventually if we ever wanted the extensibility then we would need to expose these, but that hasnt been thought through yet

Reply:
I'm not suggesting the interface needs to be publicly visible, but I do think infracopilot needs a bit more insight than just "this one constraint failed".

Reply:
yeah agreed, ill get something set up to work that in so we can produce the reasoning on the klotho engine side